runtime/storage.go

```diff
@@ -283,117 +232,117 @@ func (s *Storage) recordContractUpdate(
 
-	SortAccountStorageEntries(accountStorageEntries)
+	for _, contractUpdate := range contractUpdates {
 
-	// Write account storage entries in order
+		key := contractUpdate.Key
 
-	// TODO: bring back concurrent encoding
-	for _, entry := range accountStorageEntries {
+		storageMap := s.GetStorageMap(key.Address, StorageDomainContract)
 
-		storageKey := entry.StorageKey
-		storable := entry.Storable
+		var value interpreter.OptionalValue
 
-		address := storageKey.Address
+		if contractUpdate.Storable == nil {
+			value = interpreter.NilValue{}
+		} else {
+			contractValue := interpreter.StoredValue(contractUpdate.Storable, s)
+			value = interpreter.NewSomeValueNonCopying(contractValue)
+		}
 
-		// If the account storage change is a contract update,
-		// and it is overwriting an existing contract value,
-		// it must be properly removed first:
-		// The removal did not occur during execution,
-		// because contract updates are deferred to the commit
+		storageMap.WriteValue(inter, key.Key, value)
+	}
+}
 
-		if entry.IsContractUpdate {
-			existingStorable := s.readStorable(storageKey)
-			if existingStorable != nil {
-				interpreter.StoredValue(existingStorable, s).
-					DeepRemove(inter)
-				inter.RemoveReferencedSlab(existingStorable)
-			}
-		}
+type write struct {
+	storageKey StorageKey
+	data       []byte
+}
 
-		var encoded []byte
+func sortWrites(writes []write) {
+	sort.Slice(writes, func(i, j int) bool {
+		a := writes[i].storageKey
+		b := writes[j].storageKey
+		return a.IsLess(b)
+	})
+}
 
-		if storable != nil {
-			var err error
+// Commit serializes/saves all values in the readCache in storage (through the runtime interface).
+//
+func (s *Storage) Commit(inter *interpreter.Interpreter, commitContractUpdates bool) error {
 
-			var buf bytes.Buffer
-			encoder := atree.NewEncoder(&buf, interpreter.CBOREncMode)
+	if commitContractUpdates {
+		s.commitContractUpdates(inter)
+	}
 
-			s.reportMetric(
-				func() {
-					err = storable.Encode(encoder)
-				},
-				func(metrics Metrics, duration time.Duration) {
-					metrics.ValueEncoded(duration)
-				},
-			)
-			if err != nil {
-				return err
-			}
+	var writes []write
 
-			err = encoder.CBOR.Flush()
-			if err != nil {
-				return err
-			}
+	// NOTE: ranging over maps is safe (deterministic),
+	// if it is side effect free and the keys are sorted afterwards
+
+	for storageKey, data := range s.writes { //nolint:maprangecheck
+		writes = append(
+			writes,
+			write{
+				storageKey: storageKey,
+				data:       data,
+			},
+		)
+	}
 
-			encoded = buf.Bytes()
-		}
+	// Sort the writes by storage key in lexicographic order
+
+	sortWrites(writes)
+
+	// Write account storage entries in order
+
+	for _, write := range writes {
 
 		var err error
 		wrapPanic(func() {
 			err = s.Ledger.SetValue(
-				address[:],
-				[]byte(storageKey.Key),
-				encoded,
+				write.storageKey.Address[:],
+				[]byte(write.storageKey.Key),
+				write.data,
```

Comment:
Not sure, but if `write.data` is a storage index referring to an empty map, should we remove the entry instead of setting a value?